### PR TITLE
✨ (rc): Add Update Service

### PR DESCRIPTION
**Pull request by YannLocatelli**

### CODE REVIEW

**Review by ladislas — COMMENTED** [review]
LGTM 👍 

quelques suggestions sur le naming et des questions sur les tests.
on pourra en parler en réunion

File: libs/RobotKit/include/RobotController.h [hunk 1]
```diff
@@ -134,7 +135,7 @@ class RobotController : public interface::RobotController
 
 	void applyUpdate() final
 	{
-		auto firmware_version = FirmwareVersion {.major = 1, .minor = 2, .revision = 3};
+		auto firmware_version = _service_update.getVersion();
 		if (_firmware_update.loadUpdate(firmware_version) && _on_update_loaded_callback != nullptr) {
```
[the remark this answers is not in the document]
Comment (YannLocatelli) [reply libs/RobotKit/include/RobotController.h:139]: It is not the same callback, the one from BLEServiceUpdate `_on_apply_update_callback` is an instruction from user

`_on_update_loaded_callback` is used after the selected firmware has been transferred (loaded) into the external flash (via QSPI), to give instruction to mcuboot, which is the one below today

https://github.com/leka/LekaOS/blob/fe54bcf9fb48676617f09a24f687e6e8d3b7f964/app/os/main.cpp#L200

File: libs/BLEKit/include/BLEServiceUpdate.h [hunk 2]
```diff
@@ -35,10 +38,13 @@ class BLEServiceUpdate : public interface::BLEService
 		}
 	}
 
+	void onApplyUpdateTrue(const std::function<void()> &callback) { _on_apply_update_true = callback; }
```
[the remark this answers is not in the document]
Comment (YannLocatelli) [reply libs/BLEKit/include/BLEServiceUpdate.h]: It is called before applying the update

File: libs/BLEKit/include/BLEServiceUpdate.h [hunk 3]
```diff
@@ -35,10 +38,13 @@ class BLEServiceUpdate : public interface::BLEService
...
+
   private:
 	bool apply_update_value {false};
```
[the remark this answers is not in the document]
Comment (YannLocatelli) [reply libs/BLEKit/include/BLEServiceUpdate.h]: Variable containing the value of the characteristic.

I simplified it to be only the default value of the characteristic.

File: libs/RobotKit/include/RobotController.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ladislas) [reply libs/RobotKit/include/RobotController.h:139]: I know it's not the same callback. I meant that in one case you compare it against `nullptr` with `if (callback != nullptr) {...}` and in the other case you have implicit nullptr to bool conversion with `if (callback) {...}`

style should be consistent. After giving it second thought, I think the explicit version is clearer and future proof.

That's what google says, see: https://abseil.io/tips/141

File: libs/BLEKit/include/BLEServiceUpdate.h [hunk 4]
```diff
@@ -35,10 +36,13 @@ class BLEServiceUpdate : public interface::BLEService
 		}
 	}
 
+	void onApplyUpdate(const std::function<void()> &callback) { _on_apply_update_callback = callback; }
+
   private:
-	bool apply_update_value {false};
+	bool default_apply_update {false};
```
Comment (ladislas) [inline libs/BLEKit/include/BLEServiceUpdate.h:42]: si je comprends bien c'est la valeur par défaut? 

est-ce qu'elle peut être `const`?

en même temps elle est utilisé just en dessous dans la characteristic, du coup elle est pas const j'imagine, mais dans ce cas là, le `default` est bizarre puisque si c'est pas défaut, c'est généralement const.

File: libs/RobotKit/include/RobotController.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (YannLocatelli) [reply libs/RobotKit/include/RobotController.h:139]: If google said it

**Pull request merged**